src/Datadog.Trace.ClrProfiler.Managed/Integrations/Testing/AsyncTool.cs

```diff
@@ -0,0 +1,215 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.ClrProfiler.Integrations.Testing
+{
+    /// <summary>
+    /// This class is similar to AsyncHelper but removes the params array allocation, the string key allocation and the invoke by reflection.
+    /// </summary>
+    internal static class AsyncTool
+    {
+        private static readonly ConcurrentDictionary<Type, TaskContinuationGenerator> ContinuationsGeneratorCache = new ConcurrentDictionary<Type, TaskContinuationGenerator>();
+
+        /// <summary>
+        /// Adds a continuation based on the current returnValue
+        /// </summary>
+        /// <typeparam name="TState">Type of the state</typeparam>
+        /// <param name="returnValue">Return value</param>
+        /// <param name="ex">Exception</param>
+        /// <param name="continuation">Continuation delegate</param>
+        /// <param name="state">State value</param>
+        /// <param name="generator">Continuation generator</param>
+        /// <returns>Return value after the continuation</returns>
+        public static object AddContinuation<TState>(object returnValue, Exception ex, Func<object, Exception, TState, object> continuation, TState state, TaskContinuationGenerator generator = null)
```

Comment:
nitpick: For me the signature would be a little more readable if we put the state object before the continuation function. Do you have a strong preference either way on this signature?

Reply:
No, make sense, I'll do the modification.

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/802/commits/9ca2dae97b9b229f56788d854e75323b096a21ad